### DISCUSSION TITLE
Why did you choose this way of filtering?

### DISCUSSION THREAD

**Discussion by frederikhors** [ent/ent#2582] [Q&A] [answered]
I'm using GraphQL in Golang with [gqlgen](https://github.com/99designs/gqlgen) and [Bun](https://github.com/uptrace/bun) as ORM.

I'm evaluating ent right now.

I'm using [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) and [Clean architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html), meaning that I'm avoiding `one-size-fits-all/lock-in` solutions like Hasura, Postgraphile, DGraph and everything like that; I can (and I eventually will, no hurry) write and organize my code in such a way to achieve their features.

For now I'm wondering about complex queries with filters like: `AND`, `OR`, `NOT` ([better described here](https://github.com/graphile-contrib/postgraphile-plugin-connection-filter/blob/master/docs/examples.md)).

Right now I need to think/write/test code all by myself and this certainly does not speed up production of the software I want to write.

Also, my code will almost certainly be slow, buggy and possibly all wrong as soon as I realize that I will need something extra for this feature.

_After all, this is true for all software and is one of the reasons why open source was born._

After some research I noticed there are two approaches to solving the problem of GraphQL data filtering:

- create all possible combinations for each field of the model ([this is how ent behaves](https://entgo.io/blog/2021/07/01/automatic-graphql-filter-generation) and [other users suggest](https://github.com/99designs/gqlgen/issues/385#issuecomment-432130569))

- create input types that contain the model fields but also logical comparators as a possible value (like [Hasura, Postgraphile and others do](https://hasura.io/docs/latest/graphql/core/databases/postgres/queries/query-filters/))

### QUESTION

Why did you choose this path instead of doing like Hasura and Postgraphile?

Pros and cons of each choice?

_Amazing work with Ent, really a huge project with amazing features! Congratulations!_

**Comment by a8m** [comment, marked as answer]
Hey @frederikhors 👋 

First, thanks for the kind words! Second, let me answer this really briefly and I'll love to be more specific if something is not clear.

> Why did you choose this path instead of doing like Hasura and Postgraphile?

I went with other approaches in the past (e.g. when I built [RQL](https://github.com/a8m/rql)), but when I decided to support it in Ent, I thought that the GraphQL predicate construction should be ~identical to the same way it's done in Ent. For example:

```graphql
where: {
  hasParent: true,
  hasChildrenWith: {
    status: IN_PROGRESS,
  },
  createdAtGT: ...,
}
```

```go
Where(
    todo.HasParent(),
    todo.HasChildrenWith(
        todo.Status(todo.StatusInProgress),
    ),
    todo.CreatedAtGT(...),
)
```
Both approaches work and you can build the same `WHERE` clause with the 2 options. The question is why I went with the first approach for Ent. The reason is that it's shorter and cleaner for most cases (IMHO) - I can write here a few examples. Also, the API is much friendlier when working on editors with "auto-completion" support. 

![image](https://user-images.githubusercontent.com/7413593/171275298-72e56d9a-6cb0-4ff1-ae5d-03016dee4541.png)

> Pros and cons of each choice?

It's really easy to craft the other approach with external templates (the same way entgql filters are implemented) and see what approach works better for you. I'll be happy to support you if you'll need help with it.
